File: apps/mobile/targets/positions-widget/Sources/Rendering/Protocols/UniswapV3/RangeBarView.swift
import SwiftUI

struct RangeBarView: View {
  let range: WidgetTickRange
  let trackHeight: CGFloat
  let thumbSize: CGFloat

  init(range: WidgetTickRange, trackHeight: CGFloat = 8, thumbSize: CGFloat = 16) {
    self.range = range
    self.trackHeight = trackHeight
    self.thumbSize = thumbSize
  }

  private var bar: RangeMath.BarPositions {
    RangeMath.barPositions(
      currentTick: range.currentTick,
      tickLower: range.tickLower,
      tickUpper: range.tickUpper
    )
  }

  private var fill: Color {
    bar.inRange ? .statusInRange : .statusOutOfRange
  }

  var body: some View {
    let positions = bar
    GeometryReader { geo in
      ZStack(alignment: .leading) {
        Capsule()
          .fill(Color.textPrimary.opacity(0.10))
          .frame(height: trackHeight)
          .frame(maxWidth: .infinity, alignment: .leading)

        Capsule()
          .fill(
            LinearGradient(
              colors: [fill.opacity(0.85), fill],
              startPoint: .leading,
              endPoint: .trailing
            )
          )
          .frame(
            width: max(0, geo.size.width * positions.liquidityWidthPct),
            height: trackHeight
          )
          .offset(x: geo.size.width * positions.liquidityLeftPct)

        Circle()
          .fill(fill)
          .frame(width: thumbSize, height: thumbSize)
          .overlay(Circle().stroke(Color.bgPrimary, lineWidth: 2.5))
          .shadow(color: fill.opacity(0.55), radius: 6, x: 0, y: 0)
          .offset(x: geo.size.width * positions.thumbPct - thumbSize / 2)
      }
      .frame(height: max(thumbSize, trackHeight), alignment: .center)
    }
    .frame(height: max(thumbSize, trackHeight))
  }
}

File: apps/mobile/modules/widget-bridge/android/src/main/java/expo/modules/widgetbridge/widget/ui/components/RangeBarView.kt
package expo.modules.widgetbridge.widget.ui.components

import android.graphics.Bitmap
import android.graphics.BlurMaskFilter
import android.graphics.Canvas
import android.graphics.LinearGradient
import android.graphics.Paint
import android.graphics.Shader
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.graphics.toArgb
import androidx.glance.GlanceModifier
import androidx.glance.Image
import androidx.glance.ImageProvider
import androidx.glance.layout.fillMaxWidth
import androidx.glance.layout.height
import expo.modules.widgetbridge.widget.data.WidgetTickRange
import expo.modules.widgetbridge.widget.theme.Colors
import expo.modules.widgetbridge.widget.theme.Opacity
import expo.modules.widgetbridge.widget.theme.Sizing
import expo.modules.widgetbridge.widget.util.RangeMath
import kotlin.math.max

private const val BAR_WIDTH_PX = 600
private const val BAR_HEIGHT_PX = 48

private fun render(range: WidgetTickRange): Bitmap {
  val pos = RangeMath.barPositions(range.currentTick, range.tickLower, range.tickUpper)
  val w = BAR_WIDTH_PX.toFloat()
  val h = BAR_HEIGHT_PX.toFloat()
  val trackHeight = 40f
  val thumbDiameter = 46f
  val fillColor = if (pos.inRange) Colors.statusInRange else Colors.statusOutOfRange

  val bmp = Bitmap.createBitmap(BAR_WIDTH_PX, BAR_HEIGHT_PX, Bitmap.Config.ARGB_8888)
  val canvas = Canvas(bmp)
  val paint = Paint(Paint.ANTI_ALIAS_FLAG)

  // Inactive track
  paint.color = Colors.textPrimary.copy(alpha = Opacity.strokeStrong).toArgb()
  canvas.drawRoundRect(
    0f, (h - trackHeight) / 2,
    w, (h + trackHeight) / 2,
    trackHeight / 2, trackHeight / 2, paint,
  )

  // Active gradient capsule
  val activeLeft = (w * pos.liquidityLeftPct).toFloat()
  val activeWidth = max(0f, (w * pos.liquidityWidthPct).toFloat())
  if (activeWidth > 0) {
    val activeRight = activeLeft + activeWidth
    paint.shader = LinearGradient(
      activeLeft, 0f, activeRight, 0f,
      fillColor.copy(alpha = Opacity.gradientStart).toArgb(),
      fillColor.toArgb(),
      Shader.TileMode.CLAMP,
    )
    canvas.drawRoundRect(
      activeLeft, (h - trackHeight) / 2,
      activeRight, (h + trackHeight) / 2,
      trackHeight / 2, trackHeight / 2, paint,
    )
    paint.shader = null
  }

  // Thumb
  val thumbX = (w * pos.thumbPct).toFloat().coerceIn(thumbDiameter / 2, w - thumbDiameter / 2)
  val thumbY = h / 2

  // Glow
  paint.color = fillColor.copy(alpha = Opacity.glow).toArgb()
  paint.maskFilter = BlurMaskFilter(12f, BlurMaskFilter.Blur.NORMAL)
  canvas.drawCircle(thumbX, thumbY, thumbDiameter / 2 + 4f, paint)
  paint.maskFilter = null

  // Fill
  paint.style = Paint.Style.FILL
  paint.color = fillColor.toArgb()
  canvas.drawCircle(thumbX, thumbY, thumbDiameter / 2, paint)

  // Stroke
  paint.style = Paint.Style.STROKE
  paint.strokeWidth = Sizing.RangeBar.thumbStrokePx
  paint.color = Colors.bgPrimary.toArgb()
  canvas.drawCircle(thumbX, thumbY, thumbDiameter / 2, paint)

  return bmp
}

@Composable
fun RangeBarView(range: WidgetTickRange) {
  val bitmap = remember(range) { render(range) }
  Image(
    provider = ImageProvider(bitmap),
    contentDescription = null,
    modifier = GlanceModifier.fillMaxWidth().height(Sizing.RangeBar.thumb),
  )
}
